## Part I Warm up with the tf.keras Sequential API

### 1) Linear Model

In [0]:
import tensorflow as tf
import numpy as np

In [2]:
mnist = tf.keras.datasets.mnist

(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train, x_test = x_train / 255.0, x_test / 255.0
y_train = tf.keras.utils.to_categorical(y_train)
y_test = tf.keras.utils.to_categorical(y_test)

Linear_model = tf.keras.models.Sequential([
    tf.keras.layers.Flatten(),
    tf.keras.layers.Dense(10, activation = 'sigmoid')
])

Linear_model.compile(optimizer=tf.train.AdamOptimizer(0.0005),
              loss='categorical_crossentropy',
              metrics=['accuracy'])

11493376/11490434 [==============================] - 0s 0us/step


In [3]:
Linear_model.fit(x_train, y_train, epochs = 5)

Epoch 1/5
60000/60000 [==============================] - 2s 38us/step - loss: 0.6225 - acc: 0.8531
Epoch 2/5
60000/60000 [==============================] - 2s 37us/step - loss: 0.3399 - acc: 0.9080
Epoch 3/5
60000/60000 [==============================] - 2s 37us/step - loss: 0.3070 - acc: 0.9149
Epoch 4/5
60000/60000 [==============================] - 2s 36us/step - loss: 0.2911 - acc: 0.9196
Epoch 5/5
60000/60000 [==============================] - 2s 36us/step - loss: 0.2814 - acc: 0.9218


In [4]:
Linear_model.evaluate(x_test, y_test)

10000/10000 [==============================] - 0s 23us/step


[0.27315028735995295, 0.9241]

### 2) Deep Model 

In [0]:
Deep_model = tf.keras.models.Sequential([
    tf.keras.layers.Flatten(),
    tf.keras.layers.Dense(512, activation = 'sigmoid'),
    tf.keras.layers.Dense(512, activation = 'sigmoid'),
    tf.keras.layers.Dense(10, activation = 'sigmoid')
])

Deep_model.compile(optimizer=tf.train.AdamOptimizer(0.0005),
              loss='categorical_crossentropy',
              metrics=['accuracy'])


In [6]:
Deep_model.fit(x_train, y_train, epochs = 5)

Epoch 1/5
60000/60000 [==============================] - 15s 243us/step - loss: 0.4271 - acc: 0.8716
Epoch 2/5
60000/60000 [==============================] - 14s 240us/step - loss: 0.1903 - acc: 0.9431
Epoch 3/5
60000/60000 [==============================] - 14s 237us/step - loss: 0.1334 - acc: 0.9599
Epoch 4/5
60000/60000 [==============================] - 14s 235us/step - loss: 0.0986 - acc: 0.9700
Epoch 5/5
60000/60000 [==============================] - 14s 236us/step - loss: 0.0756 - acc: 0.9767


In [7]:
Deep_model.evaluate(x_test, y_test)

10000/10000 [==============================] - 1s 78us/step


[0.0823746071472764, 0.9746]

## Evaluation

**Linear Model**: It has one output layer with 10 nodes with the sigmoid activation function without any hidden layer. The test accuracy for this linear model is about 0.92.

**Deep model**: The deep model that I built has 2 hidden layers each with 512 nodes and sigmoid activation function. The output layer has 10 nodes with sigmoid activation function as well. The test accuracy for the deep model can reach 0.97.